### PR TITLE
Fix broken login flow

### PR DESCRIPTION
**Pull request by MrFlashAccount**
#### Tl;dr 
Closes: enso-org/cloud-v2#1231
This PR makes the ENSO_CLOUD_REDIRECT env variable optional and uses window.location.origin as a redirect url by default

---

#### Context:
In the recent PR we have changed the server address where we start the electron app from 'http' to 'https'. This change was caused by the Stripe live mode that doesn't work in the HTTP environment. But now we have different origins based on where we launched the app. So we decided to make the ENSO_CLOUD_REDIRECT env variable optional and use window.location.origin as a redirect URL by default. Also, we found a bug if the electron server started on a different port, let's say 8081(this might happen if 8080 is already in use) and the server will be implicitly redirected from 8081 to 8080 after login. We decided to explicitly kill the app and show the error in the console. This is because we need to specify allowed redirect URLs in COGNITO explicitly to make the login flow work.

#### This Change:
1. Makes ENSO_CLOUD_REDIRECT optional
2. Uses window.location.origin as a defualt redirect URL
3. Throw an error if the port is in use when starting an electron app

#### Test Plan:
Go over how you plan to test it. Your test plan should be more thorough the riskier the change is. For major changes, I like to describe how I E2E tested it and will monitor the rollout.

---

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: app/ide-desktop/lib/client/src/bin/server.ts [hunk 1]
```diff
@@ -78,6 +80,18 @@ async function findPort(port: number): Promise<number> {
...
 
+// ================
+// === Port Use ===
+// ================
+
+/**
+ * Check if a port is in use.
+ */
+async function isPortInUse(port: number): Promise<boolean> {
+    const nearestPort = await findPort(port)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/client/src/bin/server.ts]: we use this approach because (for some weird reason) approach with starting a server on the port doesn't work well(it returns ok even if the port is busy)

File: app/ide-desktop/lib/dashboard/src/authentication/service.ts [hunk 2]
```diff
@@ -167,10 +167,11 @@ function loadAmplifyConfig(
     // needs to be registered.
     setDeepLinkHandler(logger, navigate)
   }
+
+  const redirectUrl = process.env.ENSO_CLOUD_REDIRECT ?? window.location.origin
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: nit: we might also want to use `APP_BASE_URL`, to cover the rare case when the app is not hosted at `/` (e.g. it's being accessed through a proxy)

**Review by somebody1234 — APPROVED** [review]
CR passed :white_check_mark: 
QA not done (at least not yet)

File: app/ide-desktop/lib/client/src/bin/server.ts (hunk 1 above)
Comment (somebody1234) [reply app/ide-desktop/lib/client/src/bin/server.ts]: we might want to limit the amount of ports checked? not sure how difficult that is

File: app/ide-desktop/lib/client/watch.ts [hunk 3]
```diff
@@ -109,18 +109,32 @@ process.on('SIGINT', () => {
     })
 })
 
-while (true) {
+/**
+ * Starts the electron process with the IDE.
+ */
+function startElectronProcess() {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/client/watch.ts:115]: I refactored the watch script and disabled continuous restarting of the app because it leads to a worse experience(e.g. when I close the app it restarts immediately. Same happens when process fails and leads to infinite cycle of restarting app)

File: app/ide-desktop/lib/client/src/bin/server.ts (hunk 1 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/bin/server.ts]: Yeah, it's possible. Good catch

File: app/ide-desktop/lib/client/watch.ts (hunk 3 above)
Comment (somebody1234) [reply app/ide-desktop/lib/client/watch.ts:115]: yeah makes sense. i kept it around because it was a thing in the old watch script, but as both dashboard and GUI use vite, HMR should more or less take care of everything

File: app/ide-desktop/lib/dashboard/src/authentication/service.ts (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: Maybe we can use the full URL instead, wdyt?

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: Or did you mean to concat `ENSO_CLOUD_REDIRECT` with `APP_BASE_URL`?
Or both? 😄

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: that might work, yeah. note that we use path for react-router on cloud.enso.org, but if this behavior does not exist on cloud.enso.org then it's not a problem

(on second thought, it shouldn't be a problem either way right? because if we end up at an invalid path, the react router `Layout`s should take care of that. (and if they don't, then it's a bug that should be fixed anyway)

either way should probably be tested on both electron and browser to make sure nothing breaks

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: >  concat `ENSO_CLOUD_REDIRECT` with `APP_BASE_URL`

yeah this is what i was imagining. not sure whether it's necessary though (see above)

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: Yeah, as far as I remember we redirect all unknown paths to the Dashboard: https://github.com/enso-org/enso/blob/f0411766fa16782a588e1cd50c32f5ba4fccf3f0/app/ide-desktop/lib/dashboard/src/App.tsx#L450

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/authentication/service.ts:171]: Not sure it's inside the scope of the current issue, created a separate one: https://github.com/enso-org/cloud-v2/issues/1240

File: app/ide-desktop/lib/client/src/bin/server.ts [hunk 4]
```diff
@@ -98,10 +121,20 @@ export class Server {
...
+        if (await isPortInUse(localConfig.port)) {
+            logger.error(
+                `💥Port ${localConfig.port} is already in use. Please close the application using it and try again.`
+            )
+
+            return Promise.reject(
+                `💥Port ${localConfig.port} is already in use. Please close the application using it and try again.`
+            )
+        } else {
+            const server = new Server(localConfig)
```
Comment (Frizi) [inline app/ide-desktop/lib/client/src/bin/server.ts]: This is not a reliable way to check for port availability. Since the logic is all async and non-atomic, you have no guarantee that the port will stay unused between the `isPortInUse` and `server.run`. This sort of race conditions have already caused us a fair amount of failed CI runs. The better strategy is to just attempt to run a server with specified port, and check afterwards if it was successful or not. Just don't run the `findPort` logic at all.

File: app/ide-desktop/lib/client/src/index.ts [hunk 5]
```diff
@@ -223,8 +223,8 @@ class App {
             await logger.asyncGroupMeasured('Starting the application', async () => {
                 // Note that we want to do all the actions synchronously, so when the window
                 // appears, it serves the website immediately.
-                await this.startBackendIfEnabled()
                 await this.startContentServerIfEnabled()
+                await this.startBackendIfEnabled()
```
Comment (Frizi) [inline app/ide-desktop/lib/client/src/index.ts:227]: Would it be possible to run those in parallel?
```suggestion
                await Promise.all([
                    this.startContentServerIfEnabled(),
                    this.startBackendIfEnabled()
                ])
```

File: app/ide-desktop/lib/client/watch.ts [hunk 6]
```diff
@@ -109,18 +109,32 @@ process.on('SIGINT', () => {
...
+            electronProcess.removeAllListeners()
+            process.exit(0)
+        }
+    })
+    electronProcess.on('error', error => {
+        console.error('Electron process failed:', error)
+        console.log('Trying to restart Electron process.')
+
+        electronProcess.removeAllListeners()
+        startElectronProcess()
```
Comment (Frizi) [inline app/ide-desktop/lib/client/watch.ts]: This leads to an infinite loop in case of recurring startup error.

**Review by Frizi — APPROVED** [review]
Approving the vite config modification.

File: app/ide-desktop/lib/client/src/index.ts (hunk 5 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/index.ts:227]: We can, but I think if the content server fails to start, we don't need the backend. Also, the logs look prettier (the error comes last with this approach, which is minor, though). WDYT?

File: app/ide-desktop/lib/client/watch.ts (hunk 6 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/watch.ts]: Agree, will remove it

File: app/ide-desktop/lib/client/src/bin/server.ts (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/bin/server.ts]: I mentioned this approach here: https://github.com/enso-org/enso/pull/9965#discussion_r1601638151

Lemme think about how we can tackle this...

Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/bin/server.ts]: If we start the server, we're getting this error:
```
(node:65794) electron: Failed to load URL: https://localhost:8080/?engine.projectManagerUrl=ws%3A%2F%2F127.0.0.1%3A30535 with error: ERR_SSL_PROTOCOL_ERROR
```
which is, IMO absolutely uninformative and the server notified that it successfully started on the port and it is weird.

Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/bin/server.ts]: As an alternative, I can put the check later to reduce the possible time lag between checking the port and starting the server. 

Also, I think even if the logic was all sync, it can't be reliable at 100% (please let me know if I'm wrong).

On the other though, if we try to launch 2 IDE on the same machine we'd also fail in production...

Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/bin/server.ts]: Will raise this question tommorow on daily, will see, maybe we'll rewrite this part completely as it also has security issues.

File: app/ide-desktop/lib/client/src/index.ts (hunk 5 above)
Comment (Frizi) [reply app/ide-desktop/lib/client/src/index.ts:227]: Sounds fine to keep it ordered if it causes less frustration 👍

File: app/ide-desktop/lib/client/src/bin/server.ts (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/client/src/bin/server.ts]: We decided to roll back the logic back to the original (start on free next port).

**Comment by MrFlashAccount** [comment]
@somebody1234 awaiting QA

**Review by PabloBuchu — APPROVED** [review]

**Comment by PabloBuchu** [comment]
QA 🟢
